targets/esp32.ld

```diff
@@ -0,0 +1,95 @@
+/* Linker script for the ESP32 */
+
+MEMORY
+{
+    /* Data RAM. Allows byte access.
+     * There are various data RAM regions:
+     *   SRAM2: 0x3FFA_E000..0x3FFD_FFFF (72 + 128 = 200K)
+     *   SRAM1: 0x3FFE_0000..0x3FFF_FFFF (128K)
+     * This gives us 328K of contiguous RAM, which is the largest span possible.
+     * SRAM1 has other addresses as well but the datasheet seems to indicate
+     * these are aliases.
+     */
+    DRAM  (rw) : ORIGIN = 0x3FFAE000, LENGTH = 200K + 128K /* Internal SRAM 1 + 2 */
```

Comment:
One thing to watch out for is that the startup stack is somewhere in that range. Might be useful to add an assert that the .data/.rodata sections don't overlap the stack used by the bootloader.

Reply:
That's what gave me trouble initially. I solved it by setting the stack pointer at startup, see src/device/esp/esp32.S. That also makes the system more flexible, as it is no longer bound by the stack that was set by the boot ROM.

Reply:
My point is that the ROM bootloader is using certain region as stack, and if the app's .data section happens to overlap the ROM bootloader stack, the stack will get overwritten, crashing the bootloader.

Reply:
Right. I see. Thank you for explaining. In that case, I think there are two options:
* Use the stack that is used by the ROM bootloader.
* Let the application load `.data` instead of letting the ROM bootloader do that.

Is the bootloader stack documented somewhere?

Reply:
PRO CPU stack is at 0x3ffe1320 — 0x3ffe3f20, and APP CPU stack is at 0x3ffe5230 — 0x3ffe7e30
(I think these values are defined somewhere in IDF, but I can't find the link right now. They can be extracted from the ROM code ELF file, provided at https://dl.espressif.com/dl/esp32_rom.elf.)

In addition to that there are .bss and .data sections used by various ROM functions, at 
0x3ffe0000 — 0x3ffe0440, 0x3ffe3f20 — 0x3ffe4350, 0x3ffae000 — 0x3ffae6e0. If you do not plan to call ROM functions, and do not plan to call any libraries which in turn call ROM functions, you can ignore these other reserved regions.

On the ESP32-S2, the situation is a bit more sane, as the reserved regions are near the end of DRAM, and don't make "holes" in the continuous region you would otherwise use for the application.

Reply:
Thanks, that is some great insight. Also thanks for the ROM file which will be useful. I will take a look and improve this PR accordingly.
It may be possible to use the given area for the heap anyway and block out the parts that are used by the ROM code. That shouldn't need any changes to the TinyGo heap. Although I'll probably ignore that for this PR and do that only when we actually want to call these special ROM functions.

> On the ESP32-S2, the situation is a bit more sane, as the reserved regions are near the end of DRAM, and don't make "holes" in the continuous region you would otherwise use for the application.

Glad to hear! Unfortunately the ESP32 will be around for a while and we'll need to adjust to it somehow.

Reply:
> It may be possible to use the given area for the heap anyway and block out the parts that are used by the ROM code.

Right, that's what we do in IDF — these ROM stack regions are getting added to the heap allocator once the startup is complete. And the regions used for ROM .data/.bss are marked as reserved.

> Unfortunately the ESP32 will be around for a while and we'll need to adjust to it somehow.

Of course, it was more of me saying that we realize that reserved regions in the middle of DRAM are tricky to work with, and in the subsequent chips the situation is somewhat improved.